### PR TITLE
6822: Table utilites refactor

### LINKED ISSUE #6822: Table utilities files are a mess
## Provide a description of the task

Table utilities grew in time to multiple methods spread across many files. It is often random on where they land and in some times we have:

```javascript
import { findAncestor } from './utils';
import { getRowIndexes, getSelectionAffectedTableCells } from '../utils';
// plus
import { cropTableToDimensions } from './tableselection/croptable';
```
I would like to illustrate the current state and the proposal in the comments below.

### PR DESCRIPTION
**Pull request by jodator**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Other: Cleanup the table helper functions API. Closes #6822.
MINOR BREAKING CHANGE: The `isTableWidget()` and `toTableWidget()` utility functions were removed.
MINOR BREAKING CHANGE: The functions `getSelectedTableWidget()` and `getTableWidgetAncestor()` from `table/utils` module were moved to `table/utils/widget` module.
MINOR BREAKING CHANGE: The functions `getSelectedTableCells()`, `getTableCellsContainingSelection()`, `getSelectionAffectedTableCells()`, `getRowIndexes()`, `getColumnIndexes()`, and `isSelectionRectangular()` from `table/utils` module were moved to `table/utils/selection` module.
MINOR BREAKING CHANGE: The functions `getVerticallyOverlappingCells()`, `splitHorizontally()`, `getHorizontallyOverlappingCells()`, and `splitVertically()` from `table/utils` module were moved to `table/utils/structure` module.
MINOR BREAKING CHANGE: The functions `findAncestor()`, `updateNumericAttribute()`, `createEmptyTableCell()`, and `isHeadingColumnCell()` from `table/commands/utils` module were moved to `table/utils/common` module.
MINOR BREAKING CHANGE: The functions `getSingleValue()` and `addDefaultUnitToNumericValue()` from `table/commands/utils` module were moved to `table/utils/table-properties` module.
MINOR BREAKING CHANGE: The functions `cropTableToDimensions()` and `trimTableCellIfNeeded()` from `table/tableselection/croptable` module were moved to `table/utils/structure` module.
MINOR BREAKING CHANGE: The functions `repositionContextualBalloon()`, `getBalloonTablePositionData()`, and `getBalloonCellPositionData()` from `table/ui/utils` module were moved to `table/utils/ui/contextualballoon` module.
MINOR BREAKING CHANGE: The functions `getBorderStyleLabels()`, `getLocalizedColorErrorText()`, `getLocalizedLengthErrorText()`, `colorFieldValidator()`, `lengthFieldValidator()`, `lineWidthFieldValidator()`, `getBorderStyleDefinitions()`, `fillToolbar()`, and `getLabeledColorInputCreator()` from `table/ui/utils` module were moved to `table/utils/ui/table-properties` module.
MINOR BREAKING CHANGE: The constant `defaultColors` from `table/ui/utils` modules was moved to `table/utils/ui/table-properties` module.

---

### Additional information

Sorry :crying_cat_face:

### CODE REVIEW

File: packages/ckeditor5-table/tests/tablenavigation.js [hunk 1]
```diff
@@ -26,6 +25,7 @@ import { getData as getModelData, setData as setModelData } from '@ckeditor/cked
 import { assertEqualMarkup } from '@ckeditor/ckeditor5-utils/tests/_utils/utils';
 import global from '@ckeditor/ckeditor5-utils/src/dom/global';
 import env from '@ckeditor/ckeditor5-utils/src/env';
+import { getTableCellsContainingSelection } from '../src/utils/selection';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tablenavigation.js:28]: This should be moved to line 11.

File: packages/ckeditor5-table/tests/utils/structure.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
+import VirtualTestEditor from '@ckeditor/ckeditor5-core/tests/_utils/virtualtesteditor';
+import TableEditing from '../../src/tableediting';
+
+import { setData as setModelData } from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
+import { modelTable } from '../_utils/utils';
+import { getHorizontallyOverlappingCells, getVerticallyOverlappingCells } from '../../src/utils/structure';
+
+describe( 'table utils', () => {
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/utils/structure.js:14]: Missing tests for:

*   cropTableToDimensions
*   splitHorizontally
*   splitVertically
*   trimTableCellIfNeeded

File: packages/ckeditor5-table/src/utils/structure.js [hunk 3]
```diff
@@ -91,6 +91,170 @@ export function cropTableToDimensions( sourceTable, cropDimensions, writer ) {
...
+ *		   ├───┼───┤   │   ├───│
+ *		4  │ o │ p │   │   │ q │
+ *		   └───┴───┴───┴───┴───┘
+ *
+ * will return slot info for cells: "j", "f", "k".
+ *
+ * @param {module:engine/model/element~Element} table The table to check.
+ * @param {Number} overlapRow The index of the row to check.
+ * @param {Number} [startRow=0] A row to start analysis. Use it when it is known that the cells above that row will not overlap.
+ * * @returns {Array.<module:table/tablewalker~TableSlot>}
```
Comment (niegowski) [inline packages/ckeditor5-table/src/utils/structure.js:116]:
```suggestion
 * @returns {Array.<module:table/tablewalker~TableSlot>}
```

File: packages/ckeditor5-table/src/utils/structure.js [hunk 4]
```diff
@@ -91,6 +91,170 @@ export function cropTableToDimensions( sourceTable, cropDimensions, writer ) {
...
+ *		│ n │ o     │ p │ q │
+ *		└───┴───────┴───┴───┘
+ *		              ^
+ *		              Overlap column to check
+ *
+ * will return slot info for cells: "b", "e", "i".
+ *
+ * @param {module:engine/model/element~Element} table The table to check.
+ * @param {Number} overlapColumn The index of the column to check.
+ * * @returns {Array.<module:table/tablewalker~TableSlot>}
```
Comment (niegowski) [inline packages/ckeditor5-table/src/utils/structure.js:209]:
```suggestion
 * @returns {Array.<module:table/tablewalker~TableSlot>}
```

File: packages/ckeditor5-table/src/tableselection.js [hunk 5]
```diff
@@ -13,11 +13,11 @@ import first from '@ckeditor/ckeditor5-utils/src/first';
...
 import TableUtils from './tableutils';
 import MouseEventsObserver from './tableselection/mouseeventsobserver';
-import { getColumnIndexes, getRowIndexes, getSelectedTableCells, getTableCellsContainingSelection } from './utils';
-import { findAncestor } from './commands/utils';
-import { cropTableToDimensions } from './tableselection/croptable';
 
 import '../theme/tableselection.css';
+import { getColumnIndexes, getRowIndexes, getSelectedTableCells, getTableCellsContainingSelection } from './utils/selection';
+import { findAncestor } from './utils/common';
+import { cropTableToDimensions } from './utils/structure';
```
[the remark this answers is not in the document]
Comment (jodator) [reply packages/ckeditor5-table/src/tableselection.js:20]: Fixed.

File: packages/ckeditor5-table/tests/utils/structure.js (hunk 2 above)
Comment (jodator) [reply packages/ckeditor5-table/tests/utils/structure.js:14]: Not part of the PR.

**Pull request merged**